<a href="https://colab.research.google.com/github/dineshkumarsarangapani/Machine_learning_IITM/blob/main/TDS/TDS_Project_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [88]:
book_tags_url = 'https://raw.githubusercontent.com/dineshkumarsarangapani/Machine_learning_IITM/main/TDS/book_tags.csv'
books_url = 'https://raw.githubusercontent.com/dineshkumarsarangapani/Machine_learning_IITM/main/TDS/books.csv'
ratings_url = 'https://raw.githubusercontent.com/dineshkumarsarangapani/Machine_learning_IITM/main/TDS/ratings.csv'
toread_url = 'https://raw.githubusercontent.com/dineshkumarsarangapani/Machine_learning_IITM/main/TDS/toread.csv'


Q6: Which book (title) has the most number of counts of tags given by the user ie. the book with maximum user records including all tags [book_tags.csv,books.csv] ? (S)

Q8: Which is the least used tag, i.e. mapped with the lowest number of books [book_tags.csv]? (In case of more than one tag, mention the tag id with the least numerical value) (N)

Q11: How many unique tags are there in the dataset [book_tags.csv] ? (N)

Q7: Which book (goodreads_book_id) is marked as to-read by most users [books.csv,toread.csv] ? (N)

Q1: How many books do not have an original title [books.csv] ? (N)

Q3: How many unique users are present in the dataset [ratings.csv] ? (N)

Q12: What is the mean value of rating of all the books in the dataset based on (average_rating) [books.csv] ? (F)

Q5: Which (tag_id) is the most frequently used ie. mapped with the highest number of books [book_tags.csv] ? (In case of more than one tag, mention the tag id with the least numerical value) (N)

Q2: How many unique books are present in the dataset ? Evaluate based on the 'book_id'? [books.csv] (N)

Q14: Predict sentiment using Textblob. How many positive titles (original_title) are there [books.csv] ? (cut-off >0) N

In [89]:

import pandas as pd

books_data = pd.read_csv(books_url)

In [90]:
books_data.head()

book_id  goodreads_book_id  best_book_id  work_id  books_count        isbn  \
0        2                  3             3  4640799          491   439554934   
1        4               2657          2657  3275794          487    61120081   
2       15              48855         48855  3532896          710   553296981   
3       18                  5             5  2402163          376  043965548X   
4       19                 34            34  3204327          566   618346252   

         isbn13                                                 authors  \
0  9.780440e+12                             J.K. Rowling, Mary GrandPré   
1  9.780061e+12                                              Harper Lee   
2  9.780553e+12  Anne Frank, Eleanor Roosevelt, B.M. Mooyaart-Doubleday   
3  9.780440e+12                 J.K. Rowling, Mary GrandPré, Rufus Beck   
4  9.780618e+12                                          J.R.R. Tolkien   

   original_publication_year  \
0                     1997.0   
1                     1960.0   
2                     1947.0   
3                     1999.0   
4                     1954.0   

                                                  original_title  ...  \
0                       Harry Potter and the Philosopher's Stone  ...   
1                                          To Kill a Mockingbird  ...   
2  Het Achterhuis: Dagboekbrieven 14 juni 1942 - 1 augustus 1944  ...   
3                       Harry Potter and the Prisoner of Azkaban  ...   
4                                     The Fellowship of the Ring  ...   

  ratings_count work_ratings_count  work_text_reviews_count  ratings_1  \
0       4602479            4800065                    75867      75504   
1       3198671            3340896                    72586      60427   
2       1972666            2024493                    20825      45225   
3       1832823            1969375                    36099       6716   
4       1766803            1832541                    15333      38031   

   ratings_2  ratings_3  ratings_4  ratings_5  \
0     101676     455024    1156318    3011543   
1     117415     446835    1001952    1714267   
2      91270     355756     656870     875372   
3      20413     166129     509447    1266670   
4      55862     202332     493922    1042394   

                                                  image_url  \
0      https://images.gr-assets.com/books/1474154022m/3.jpg   
1   https://images.gr-assets.com/books/1361975680m/2657.jpg   
2  https://images.gr-assets.com/books/1358276407m/48855.jpg   
3      https://images.gr-assets.com/books/1499277281m/5.jpg   
4     https://images.gr-assets.com/books/1298411339m/34.jpg   

                                            small_image_url  
0      https://images.gr-assets.com/books/1474154022s/3.jpg  
1   https://images.gr-assets.com/books/1361975680s/2657.jpg  
2  https://images.gr-assets.com/books/1358276407s/48855.jpg  
3      https://images.gr-assets.com/books/1499277281s/5.jpg  
4     https://images.gr-assets.com/books/1298411339s/34.jpg  

[5 rows x 23 columns]

In [91]:
len(books_data['book_id'].unique())

1418

#How many unique books are present in the dataset ? Evaluate based on the 'book_id'? [books.csv] (N) 
## 1418

In [92]:
#Q8: Which is the least used tag, i.e. mapped with the lowest number of books [book_tags.csv]? (In case of more than one tag, mention the tag id with the least numerical value) (N)

book_tags = pd.read_csv(book_tags_url)
book_tags.head()

goodreads_book_id  tag_id   count
0                  1   30574  167697
1                  1   11305   37174
2                  1   11557   34173
3                  1    8717   12986
4                  1   33114   12716

In [93]:
tag_count = book_tags['tag_id'].value_counts()
tag_count_leasts = tag_count[tag_count==1]
tag_count_leasts

155      1
20539    1
19168    1
30286    1
28491    1
        ..
28777    1
20601    1
7892     1
22286    1
5361     1
Name: tag_id, Length: 3897, dtype: int64

In [94]:
tag_count_leasts.sort_index()

52       1
53       1
56       1
57       1
59       1
        ..
34226    1
34230    1
34231    1
34232    1
34243    1
Name: tag_id, Length: 3897, dtype: int64

#Q8: Which is the least used tag, i.e. mapped with the lowest number of books [book_tags.csv]? (In case of more than one tag, mention the tag id with the least numerical value) (N)
## Ans: 52

# Q5: Which (tag_id) is the most frequently used ie. mapped with the highest number of books [book_tags.csv] ? (In case of more than one tag, mention the tag id with the least numerical value) (N)

## Ans: 6828

In [95]:
tag_count

6828     1418
30574    1417
11557    1402
11743    1398
8717     1396
         ... 
28777       1
20601       1
7892        1
22286       1
5361        1
Name: tag_id, Length: 7846, dtype: int64

# Q11: How many unique tags are there in the dataset [book_tags.csv] ? (N)
## Ans: 7846

In [96]:
book_tags['tag_id'].unique()

array([30574, 11305, 11557, ...,  9452,   329,  5361])

In [97]:
len(book_tags['tag_id'].unique())

7846

# Q1: How many books do not have an original title [books.csv] ? (N)

## Ans: 65

In [98]:
books_data.columns

Index(['book_id', 'goodreads_book_id', 'best_book_id', 'work_id',
       'books_count', 'isbn', 'isbn13', 'authors', 'original_publication_year',
       'original_title', 'title', 'language_code', 'average_rating',
       'ratings_count', 'work_ratings_count', 'work_text_reviews_count',
       'ratings_1', 'ratings_2', 'ratings_3', 'ratings_4', 'ratings_5',
       'image_url', 'small_image_url'],
      dtype='object')

In [99]:
books_data.isnull().sum()

book_id                        0
goodreads_book_id              0
best_book_id                   0
work_id                        0
books_count                    0
isbn                          21
isbn13                        14
authors                        0
original_publication_year      6
original_title                65
title                          0
language_code                256
average_rating                 0
ratings_count                  0
work_ratings_count             0
work_text_reviews_count        0
ratings_1                      0
ratings_2                      0
ratings_3                      0
ratings_4                      0
ratings_5                      0
image_url                      0
small_image_url                0
dtype: int64

# Q3: How many unique users are present in the dataset [ratings.csv] ? (N)

### 51409

In [100]:
ratings = pd.read_csv(ratings_url)

In [101]:
ratings.head()

user_id  book_id  rating
0        2     9296       5
1        2     3753       5
2        4       70       4
3        4      388       4
4        4       18       5

In [102]:
len(ratings['user_id'].unique())

51409

# Q12: What is the mean value of rating of all the books in the dataset based on (average_rating) [books.csv] ? (F)

## Ans: 4.072644569816643


In [103]:
books_data.columns

Index(['book_id', 'goodreads_book_id', 'best_book_id', 'work_id',
       'books_count', 'isbn', 'isbn13', 'authors', 'original_publication_year',
       'original_title', 'title', 'language_code', 'average_rating',
       'ratings_count', 'work_ratings_count', 'work_text_reviews_count',
       'ratings_1', 'ratings_2', 'ratings_3', 'ratings_4', 'ratings_5',
       'image_url', 'small_image_url'],
      dtype='object')

In [104]:
books_data['average_rating'].mean()

4.072644569816643

# Q6: Which book (title) has the most number of counts of tags given by the user ie. the book with maximum user records including all tags [book_tags.csv,books.csv] ? (S)

## Ans: Harry Potter and the Sorcerer's Stone (Harry Potter, #1)

In [105]:
books_data.head()

book_id  goodreads_book_id  best_book_id  work_id  books_count        isbn  \
0        2                  3             3  4640799          491   439554934   
1        4               2657          2657  3275794          487    61120081   
2       15              48855         48855  3532896          710   553296981   
3       18                  5             5  2402163          376  043965548X   
4       19                 34            34  3204327          566   618346252   

         isbn13                                                 authors  \
0  9.780440e+12                             J.K. Rowling, Mary GrandPré   
1  9.780061e+12                                              Harper Lee   
2  9.780553e+12  Anne Frank, Eleanor Roosevelt, B.M. Mooyaart-Doubleday   
3  9.780440e+12                 J.K. Rowling, Mary GrandPré, Rufus Beck   
4  9.780618e+12                                          J.R.R. Tolkien   

   original_publication_year  \
0                     1997.0   
1                     1960.0   
2                     1947.0   
3                     1999.0   
4                     1954.0   

                                                  original_title  ...  \
0                       Harry Potter and the Philosopher's Stone  ...   
1                                          To Kill a Mockingbird  ...   
2  Het Achterhuis: Dagboekbrieven 14 juni 1942 - 1 augustus 1944  ...   
3                       Harry Potter and the Prisoner of Azkaban  ...   
4                                     The Fellowship of the Ring  ...   

  ratings_count work_ratings_count  work_text_reviews_count  ratings_1  \
0       4602479            4800065                    75867      75504   
1       3198671            3340896                    72586      60427   
2       1972666            2024493                    20825      45225   
3       1832823            1969375                    36099       6716   
4       1766803            1832541                    15333      38031   

   ratings_2  ratings_3  ratings_4  ratings_5  \
0     101676     455024    1156318    3011543   
1     117415     446835    1001952    1714267   
2      91270     355756     656870     875372   
3      20413     166129     509447    1266670   
4      55862     202332     493922    1042394   

                                                  image_url  \
0      https://images.gr-assets.com/books/1474154022m/3.jpg   
1   https://images.gr-assets.com/books/1361975680m/2657.jpg   
2  https://images.gr-assets.com/books/1358276407m/48855.jpg   
3      https://images.gr-assets.com/books/1499277281m/5.jpg   
4     https://images.gr-assets.com/books/1298411339m/34.jpg   

                                            small_image_url  
0      https://images.gr-assets.com/books/1474154022s/3.jpg  
1   https://images.gr-assets.com/books/1361975680s/2657.jpg  
2  https://images.gr-assets.com/books/1358276407s/48855.jpg  
3      https://images.gr-assets.com/books/1499277281s/5.jpg  
4     https://images.gr-assets.com/books/1298411339s/34.jpg  

[5 rows x 23 columns]

In [106]:
books_data.columns

Index(['book_id', 'goodreads_book_id', 'best_book_id', 'work_id',
       'books_count', 'isbn', 'isbn13', 'authors', 'original_publication_year',
       'original_title', 'title', 'language_code', 'average_rating',
       'ratings_count', 'work_ratings_count', 'work_text_reviews_count',
       'ratings_1', 'ratings_2', 'ratings_3', 'ratings_4', 'ratings_5',
       'image_url', 'small_image_url'],
      dtype='object')

In [107]:
book_tags.head()

goodreads_book_id  tag_id   count
0                  1   30574  167697
1                  1   11305   37174
2                  1   11557   34173
3                  1    8717   12986
4                  1   33114   12716

In [108]:
book_tags['count'].max()

496107

In [109]:
book_tags[book_tags['count'] == 496107]

goodreads_book_id  tag_id   count
100                  3   30574  496107

In [110]:
# pd.set_option('max_rows', 99999)
pd.set_option('max_colwidth', 400)
# pd.describe_option('max_colwidth')

In [111]:
print(books_data[books_data['goodreads_book_id'] == 3]['title'])

0    Harry Potter and the Sorcerer's Stone (Harry Potter, #1)
Name: title, dtype: object


# Q7: Which book (goodreads_book_id) is marked as to-read by most users [books.csv,toread.csv] ? (N)

## Ans: 2657


In [112]:
toread = pd.read_csv(toread_url)

In [113]:
toread.head()

user_id  book_id
0      207       25
1      242       28
2      162      345
3      301     2047
4      418       42

In [114]:
toread['book_id'].value_counts()

4       1478
89      1458
265     1180
277     1160
28      1148
        ... 
4812       1
6345       1
9239       1
3936       1
9724       1
Name: book_id, Length: 1409, dtype: int64

In [115]:
len(toread[toread['book_id'] == 4])

1478

In [116]:
books_data[books_data['book_id'] == 4]

book_id  goodreads_book_id  best_book_id  work_id  books_count      isbn  \
1        4               2657          2657  3275794          487  61120081   

         isbn13     authors  original_publication_year         original_title  \
1  9.780061e+12  Harper Lee                     1960.0  To Kill a Mockingbird   

   ... ratings_count work_ratings_count  work_text_reviews_count  ratings_1  \
1  ...       3198671            3340896                    72586      60427   

   ratings_2  ratings_3  ratings_4  ratings_5  \
1     117415     446835    1001952    1714267   

                                                 image_url  \
1  https://images.gr-assets.com/books/1361975680m/2657.jpg   

                                           small_image_url  
1  https://images.gr-assets.com/books/1361975680s/2657.jpg  

[1 rows x 23 columns]

# Q2: How many unique books are present in the dataset ? Evaluate based on the 'book_id'? [books.csv] (N)

## Ans: 1418

In [117]:
len(books_data['book_id'].unique())

1418

# Q14: Predict sentiment using Textblob. How many positive titles (original_title) are there [books.csv] ? (cut-off >0) N

## Ans: 116

In [118]:
from textblob import TextBlob

In [119]:
books_data_notna = books_data.dropna()
books_data_notna.head()

book_id  goodreads_book_id  best_book_id  work_id  books_count        isbn  \
0        2                  3             3  4640799          491   439554934   
1        4               2657          2657  3275794          487    61120081   
2       15              48855         48855  3532896          710   553296981   
3       18                  5             5  2402163          376  043965548X   
4       19                 34            34  3204327          566   618346252   

         isbn13                                                 authors  \
0  9.780440e+12                             J.K. Rowling, Mary GrandPré   
1  9.780061e+12                                              Harper Lee   
2  9.780553e+12  Anne Frank, Eleanor Roosevelt, B.M. Mooyaart-Doubleday   
3  9.780440e+12                 J.K. Rowling, Mary GrandPré, Rufus Beck   
4  9.780618e+12                                          J.R.R. Tolkien   

   original_publication_year  \
0                     1997.0   
1                     1960.0   
2                     1947.0   
3                     1999.0   
4                     1954.0   

                                                  original_title  ...  \
0                       Harry Potter and the Philosopher's Stone  ...   
1                                          To Kill a Mockingbird  ...   
2  Het Achterhuis: Dagboekbrieven 14 juni 1942 - 1 augustus 1944  ...   
3                       Harry Potter and the Prisoner of Azkaban  ...   
4                                     The Fellowship of the Ring  ...   

  ratings_count work_ratings_count  work_text_reviews_count  ratings_1  \
0       4602479            4800065                    75867      75504   
1       3198671            3340896                    72586      60427   
2       1972666            2024493                    20825      45225   
3       1832823            1969375                    36099       6716   
4       1766803            1832541                    15333      38031   

   ratings_2  ratings_3  ratings_4  ratings_5  \
0     101676     455024    1156318    3011543   
1     117415     446835    1001952    1714267   
2      91270     355756     656870     875372   
3      20413     166129     509447    1266670   
4      55862     202332     493922    1042394   

                                                  image_url  \
0      https://images.gr-assets.com/books/1474154022m/3.jpg   
1   https://images.gr-assets.com/books/1361975680m/2657.jpg   
2  https://images.gr-assets.com/books/1358276407m/48855.jpg   
3      https://images.gr-assets.com/books/1499277281m/5.jpg   
4     https://images.gr-assets.com/books/1298411339m/34.jpg   

                                            small_image_url  
0      https://images.gr-assets.com/books/1474154022s/3.jpg  
1   https://images.gr-assets.com/books/1361975680s/2657.jpg  
2  https://images.gr-assets.com/books/1358276407s/48855.jpg  
3      https://images.gr-assets.com/books/1499277281s/5.jpg  
4     https://images.gr-assets.com/books/1298411339s/34.jpg  

[5 rows x 23 columns]

In [120]:
title = pd.DataFrame()
title['original_title'] = books_data_notna['original_title']
title['original_title_Subjectivity'] = books_data_notna['original_title'].apply(lambda x: TextBlob(x).sentiment.subjectivity)
title['original_title_Polarity'] = books_data_notna['original_title'].apply(lambda x: TextBlob(x).sentiment.polarity)


In [121]:
title

original_title  \
0                          Harry Potter and the Philosopher's Stone   
1                                             To Kill a Mockingbird   
2     Het Achterhuis: Dagboekbrieven 14 juni 1942 - 1 augustus 1944   
3                          Harry Potter and the Prisoner of Azkaban   
4                                        The Fellowship of the Ring   
...                                                             ...   
1409                                                   Горе от ума    
1411                                                 Divine Madness   
1413                                                  City of Masks   
1414                                                          Parts   
1417                      The Everafter War (The Sisters Grimm, #7)   

      original_title_Subjectivity  original_title_Polarity  
0                             0.0                      0.0  
1                             0.0                      0.0  
2                             0.0                      0.0  
3                             0.0                      0.0  
4                             0.0                      0.0  
...                           ...                      ...  
1409                          0.0                      0.0  
1411                          0.0                      0.0  
1413                          0.0                      0.0  
1414                          0.0                      0.0  
1417                          0.0                      0.0  

[1092 rows x 3 columns]

In [122]:
title[title['original_title_Polarity'] >0 ]

original_title  \
27                             Where the Wild Things Are   
60                                  A Light in the Attic   
65                           The Very Hungry Caterpillar   
80                                  The Call of the Wild   
89                       Complete Harry Potter Boxed Set   
...                                                  ...   
1361                           Marvelous Land of Oz, The   
1375                         First Term at Malory Towers   
1377  H.I.V.E.: Higher Institute of Villainous Education   
1395                The Berenstain Bears and Too Much TV   
1396                                             Perfect   

      original_title_Subjectivity  original_title_Polarity  
27                       0.400000                     0.10  
60                       0.700000                     0.40  
65                       0.300000                     0.20  
80                       0.400000                     0.10  
89                       0.400000                     0.10  
...                           ...                      ...  
1361                     1.000000                     1.00  
1375                     0.333333                     0.25  
1377                     0.500000                     0.25  
1395                     0.200000                     0.20  
1396                     1.000000                     1.00  

[116 rows x 3 columns]